sklearn/datasets/_arff_parser.py

```diff
@@ -265,4 +301,174 @@ def _liac_arff_parser(
         elif y.shape[1] == 0:
             y = None
 
-    return X, y, frame, nominal_attributes
+    if output_arrays_type == "pandas":
+        return X, y, frame, None
+    return X, y, None, categories
+
+
+def _pandas_arff_parser(
+    gzip_file,
+    output_type,
+    openml_columns_info,
+    feature_names_to_select,
+    target_names_to_select,
+):
+    """ARFF parser using `pandas.read_csv`.
+
+    This parser uses the metadata fetched directly from OpenML and skips the metadata
+    headers of ARFF file itself. The data is loaded as a CSV file.
+
+    Parameters
+    ----------
+    gzip_file : GzipFile instance
+        The GZip compressed file with the ARFF formatted payload.
+
+    output_type : {"numpy", "sparse", "pandas"}
+        The type of the arrays that will be returned. The possibilities are:
+
+        - `"numpy"`: both `X` and `y` will be NumPy arrays;
+        - `"sparse"`: `X` will be sparse matrix and `y` will be a NumPy array;
+        - `"pandas"`: `X` will be a pandas DataFrame and `y` will be either a
+          pandas Series or DataFrame.
+
+    openml_columns_info : dict
+        The information provided by OpenML regarding the columns of the ARFF
+        file.
+
+    feature_names_to_select : list of str
+        A list of the feature names to be selected to build `X`.
+
+    target_names_to_select : list of str
+        A list of the target names to be selected to build `y`.
+
+    Returns
+    -------
+    X : {ndarray, sparse matrix, dataframe}
+        The data matrix.
+
+    y : {ndarray, dataframe, series}
+        The target.
+
+    frame : dataframe or None
+        A dataframe containing both `X` and `y`. `None` if
+        `output_array_type != "pandas"`.
+
+    categories : list of str or None
+        The names of the features that are categorical. `None` if
+        `output_array_type == "pandas"`.
+    """
+    import pandas as pd
+
+    # read the file until the data section to skip the ARFF metadata headers
+    for line in gzip_file:
+        if line.decode("utf-8").lower().startswith("@data"):
+            break
+
+    # ARFF represents missing values with "?"
+    frame = pd.read_csv(
+        gzip_file,
+        header=None,
+        na_values=["?"],  # missing values are represented by `?`
+        comment="%",  # skip line starting by `%` since they are comments
+    )
```

Comment:
```suggestion
        comment="%",  # skip line starting by `%` since they are comments
        names=[name for name in openml_columns_info]
    )
```

Reply:
why do you prefer to do that instead of letting pandas inferring the column names while reading?

Reply:
The line right below the `read_csv` overrides the column names with `openml_columns_info`:

```python
frame.columns = [name for name in openml_columns_info]
```

Reply:
Oh right.